### ISSUE TITLE
build zopfli on macos

### ISSUE THREAD

**Issue by kosta** [NixOS/nixpkgs#32171]
## Issue description

Zopfli does not build on macos: `platforms = platforms.linux;`

### Steps to reproduce

On MacOS:

`$ nix-env -iA zopfli
error: attribute ‘zopfli’ in selection path ‘zopfli’ not found`

in `nixpkgs/pkgs/tools/compression/zopfli`:

    nix-build -E '(import <nixpkgs>{}).callPackage ./default.nix {}' --pure
    error: Package ‘zopfli-1.0.1’ in /Users/kosta/code/3rd/nixpkgs/pkgs/tools/compression/zopfli/default.nix:49 is not supported on ‘x86_64-darwin’, refusing to evaluate.

changing the platforms line to `platforms = platforms.unix` gives a build error:
`/nix/store/ab2iwb8ks5day43sw0580hdf54nfa4pv-bash-4.4-p12/bin/bash: gcc: command not found`

### Steps to fix ;)

(Sorry if I'm being stupid here, these are kinda my first steps with nix)

The trick here is that if you actually use `gcc`, it won't work - but if you use `clang` to pretend to be gcc it will work! (you're on macos, you know)

As proof, I added a link from `gcc` to `which clang` and then it fails with a linker errors for the "lib*" targets:

`ld: unknown option: -soname`

So you just need to patch the Makefile to replace `-soname` with `-install_name` and it works!

...except for apparently not being able to find the C++ std lib:

    src/zopflipng/lodepng/lodepng.h:32:10: fatal error: 'vector' file not found
    #include <vector>
             ^~~~~~~~
    1 error generated.

(but at least it works outside of nix-shell, so I hope this is actually the last issue?)

### Advice needed

1. What is the best way to provide a `gcc` and `g++` to the zopfli build that is actually clang? Is there an established technique for this?
1.1. Alternatively, is it ok to patch the Makefile to use clang directly?

2. What is the best way to replace `-soname` with -`install-name` on macos? Patching or is there something better?

3. How to fix the missing C++ stdlib? I guess this issue should just disappear if 1. is fixes properly.

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

    $ nix-shell -p nix-info --run "nix-info -m"
    error: undefined variable ‘nix-info’ at (string):1:66
    (use ‘--show-trace’ to show detailed location information)`

(seems unrelated, though)

Edit: Fixed line breaks in monospace text

**Comment by yegortimoshenko** [comment]
> What is the best way to provide a gcc and g++ to the zopfli build that is actually clang? Is there an established technique for this?

```
  postPatch = ''
    substituteInPlace Makefile \
      --replace gcc cc --replace g++ c++
  '';
```

**Comment by kosta** [comment]
Thank you, this is awesome!

This works on macos, I'll adjust this to run only on macos and send a PR:

    postPatch = ''
        substituteInPlace Makefile --replace gcc cc --replace g++ c++ --replace -soname -install_name
      '';

**Comment by yegortimoshenko** [comment]
@kosta `gcc -> cc` and `g++ -> c++` should run on Linux as well, to allow the package to be built via clang. It is agnostic way to reference C compiler and C++ compiler, respectively.

```
  postPatch = ''
    substituteInPlace Makefile --replace gcc cc --replace g++ c++
  '' + stdenv.lib.optionalString stdenv.isDarwin ''
    substituteInPlace Makefile --replace -soname -install_name
  '';
```

**Comment by kosta** [comment]
Thank you, this is awesome!

This works on macos, I'll adjust this to run only on macos and send a PR:

    postPatch = ''
        substituteInPlace Makefile --replace gcc cc --replace g++ c++ --replace -soname -install_name
      '';

**Comment by kosta** [comment]
Ok I will adjust my code / PR. Thanks!

**Comment by kosta** [comment]
Send PR #32176

**Comment by kosta** [comment]
Fixed by #32176
